### ISSUE TITLE
Java14 Full Records Support Check Validation: FinalClassCheck

### ISSUE THREAD

**Issue by nmancus1** [checkstyle/checkstyle#8480]
Child of #8452
Check documentation: https://checkstyle.sourceforge.io/config_design.html#FinalClass

From check documentation:
_Checks that a class which has only private constructors is declared as final. Doesn't check for classes nested in interfaces or annotations, as they are always final there._

[From JEP:](https://openjdk.java.net/jeps/384)
_A record is implicitly final, and cannot be abstract. These restrictions emphasize that the API of a record is defined solely by its state description, and cannot be enhanced later by another class or record._

So, this check does not apply, but let's check for false positives:

This will not compile, so no need to add compact constructors to this check:

```
➜  full-record-grammar cat TestClass.java 
 public class TestClass{
    public record MyRecord1() {
        private MyRecord1{}
    }
}

➜  full-record-grammar /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java
TestClass.java:3: error: invalid compact constructor in record MyRecord1
        private MyRecord1{}
                ^
  (canonical constructor must be public)
Note: TestClass.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
1 error

```
Also, compact constructor is always public. So, we need to check canonical constructors:
```
➜  full-record-grammar /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java
Note: TestClass.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
➜  full-record-grammar cat config.xml    
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
      <module name="FinalClass"/>
  </module>
</module>
➜  full-record-grammar cat TestClass.java
 public class TestClass{
    public record MyRecord1() {
        private MyRecord1(Object o){this();}
    }
}


➜  full-record-grammar java $RUN_LOCALE -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.35-SNAPSHOT-all.jar -c config.xml TestClass.java
Starting audit...
[ERROR] /home/nick/Desktop/full-record-grammar/TestClass.java:1:2: Class TestClass should be declared as final. [FinalClass]
Audit done.
Checkstyle ends with 1 errors.

```
We can see that the only violation is caused by the outer class, as expected.  This check still works as intended, so I think that we can close this issue.
